feat(init,manager,slice-machine): convert composite and legacy slices flow (DT-1660)

lihbr opened the pull request:
## Context

We want to enable users to upgrade old composite/legacy slices to shared slices managed with Slice Machine, allowing us to move toward the online custom type builder full retirement.

## The Solution

### Init

The init command now downloads remote custom types and creates them locally so they can be managed by Slice Machine if:
1. The repository selected is an existing repository
2. There's no custom type already available locally

#### Checklist

- [x] Implementation
- [x] Test

### Manager

The slice manager now has a new method `convertLegacySliceToSharedSlice()`. It takes in the composite slice model, information about the source and destination of this slice for the conversion process to happen.

#### Checklist

- [x] Implementation
- [x] Test

### UI

The UI now display a button to convert composite slices to shared slices on the custom type builder. It opens up a model guiding users through this process.

![Animation](https://github.com/prismicio/slice-machine/assets/25330882/4a56473e-2696-4abe-b28d-2570e7b87848)

#### Checklist

- [x] Logic
- [x] Tracking
   A new `SliceMachine Legacy Slice Converted` event is sent to Segment with the following string properties: `id`, `variation`, `library`, `conversionType`. It needs to be configured there.
- [x] Design
- [x] Copy

## Impact / Dependencies

<!--
List all the impacts your development might have on internal and external features.
Ideally this should have been discussed prior to this development.

Link of any other PRs that are related to this development.
They should also be referenced in the ticket for reviews.
-->


Should we disable [@typescript-eslint/strict-boolean-expressions](https://typescript-eslint.io/rules/strict-boolean-expressions/) ? Do we have a reason for this rule to be on? I find it really unintuitive to be forced to convert `undefined` to `false` in conditions...

## Checklist before requesting a review
- [x] I hereby declare my code ready for review.
- [x] If it is a critical feature, I have added tests.
- [x] The CI is successful.
- [x] If there could backward compatibility issues, it has been discussed and planned.




## Preview

<!--
Visual aid of your work.
It can either be screenshots or a video.
This could help reviewers to get context quickly.
-->

https://www.loom.com/share/1106f12be08f452fa9d555348a9438fc?sid=83e2ee93-03eb-4ec2-8810-78e9d730792e


<!--
A funny animal picture is welcome to close your PR!
You can find one here https://unsplash.com/s/photos/funny-animal-picture
-->

lihbr committed:
chore(e2e): setup project

lihbr committed:
feat(init): pull custom types on existing repository

lihbr committed:
feat(slice-machine): convert legacy slice modal (wip)

lihbr committed:
chore(deps): use alpha of internal types

lihbr committed:
feat(slice-machine): convert legacy slice modal

lihbr committed:
feat(manager): `convertLegacySliceToSharedSlice` method

lihbr committed:
test(manager): `convertLegacySliceToSharedSlice`

lihbr committed:
Merge branch 'master' of github.com:prismicio/slice-machine into lh/convert-legacy-slices

lihbr committed:
refactor(slice-machine): don't rely on `window.location.reload`

lihbr committed:
feat(slice-machine): track slice conversion

lihbr committed:
chore(slice-machine): typo

lihbr committed:
Merge branch 'master' of github.com:prismicio/slice-machine into lh/convert-legacy-slices

lihbr committed:
refactor(slice-machine): convert legacy slice modal form

lihbr committed:
refactor(slice-machine): convert legacy slice modal form

lihbr committed:
ci: release alpha

lihbr committed:
Publish

 - cimsirp@1.14.1-lh/convert-legacy-slices.0
 - next-upgrade@0.1.1-lh/convert-legacy-slices.0
 - sveltekit@1.14.1-lh/convert-legacy-slices.0
 - @slicemachine/adapter-next@0.3.17-lh/convert-legacy-slices.0
 - @slicemachine/adapter-nuxt@0.3.17-lh/convert-legacy-slices.0
 - @slicemachine/adapter-nuxt2@0.3.17-lh/convert-legacy-slices.0
 - @slicemachine/adapter-sveltekit@0.3.17-lh/convert-legacy-slices.0
 - @slicemachine/init@2.6.4-lh/convert-legacy-slices.0
 - @slicemachine/manager@0.11.3-lh/convert-legacy-slices.0
 - @slicemachine/plugin-kit@0.4.17-lh/convert-legacy-slices.0
 - slice-machine-ui@1.14.1-lh/convert-legacy-slices.0
 - start-slicemachine@0.11.7-lh/convert-legacy-slices.0

lihbr committed:
ci: release alpha

lihbr committed:
Merge branch 'lh/convert-legacy-slices' of github.com:prismicio/slice-machine into lh/convert-legacy-slices

lihbr committed:
Publish

 - cimsirp@null
 - next-upgrade@null
 - sveltekit@null
 - @slicemachine/adapter-next@null
 - @slicemachine/adapter-nuxt@null
 - @slicemachine/adapter-nuxt2@null
 - @slicemachine/adapter-sveltekit@null
 - @slicemachine/init@null
 - @slicemachine/manager@null
 - @slicemachine/plugin-kit@null
 - slice-machine-ui@null
 - start-slicemachine@null

lihbr committed:
ci: release alpha

lihbr committed:
Publish

 - cimsirp@1.14.1-dev-upgrade-legacy.0
 - next-upgrade@1.14.1-dev-upgrade-legacy.0
 - sveltekit@1.14.1-dev-upgrade-legacy.0
 - @slicemachine/adapter-next@0.3.17-dev-upgrade-legacy.0
 - @slicemachine/adapter-nuxt@0.3.17-dev-upgrade-legacy.0
 - @slicemachine/adapter-nuxt2@0.3.17-dev-upgrade-legacy.0
 - @slicemachine/adapter-sveltekit@0.3.17-dev-upgrade-legacy.0
 - @slicemachine/init@2.6.4-dev-upgrade-legacy.0
 - @slicemachine/manager@0.11.3-dev-upgrade-legacy.0
 - @slicemachine/plugin-kit@0.4.17-dev-upgrade-legacy.0
 - slice-machine-ui@1.14.1-dev-upgrade-legacy.0
 - start-slicemachine@0.11.7-dev-upgrade-legacy.0

lihbr committed:
feat(slice-machine): upgrade copy

lihbr committed:
Publish

 - cimsirp@1.14.1-dev-upgrade-legacy.1
 - next-upgrade@1.14.1-dev-upgrade-legacy.1
 - sveltekit@1.14.1-dev-upgrade-legacy.1
 - @slicemachine/adapter-next@0.3.17-dev-upgrade-legacy.1
 - @slicemachine/adapter-nuxt@0.3.17-dev-upgrade-legacy.1
 - @slicemachine/adapter-nuxt2@0.3.17-dev-upgrade-legacy.1
 - @slicemachine/adapter-sveltekit@0.3.17-dev-upgrade-legacy.1
 - @slicemachine/init@2.6.4-dev-upgrade-legacy.1
 - @slicemachine/manager@0.11.3-dev-upgrade-legacy.1
 - @slicemachine/plugin-kit@0.4.17-dev-upgrade-legacy.1
 - slice-machine-ui@1.14.1-dev-upgrade-legacy.1
 - start-slicemachine@0.11.7-dev-upgrade-legacy.1

lihbr commented:
- [ ] Remove CI edits before merging

lihbr committed:
feat(slice-machine): ui toasters on success

lihbr committed:
Publish

 - cimsirp@1.14.1-dev-upgrade-legacy.2
 - next-upgrade@1.14.1-dev-upgrade-legacy.2
 - sveltekit@1.14.1-dev-upgrade-legacy.2
 - @slicemachine/adapter-next@0.3.17-dev-upgrade-legacy.2
 - @slicemachine/adapter-nuxt@0.3.17-dev-upgrade-legacy.2
 - @slicemachine/adapter-nuxt2@0.3.17-dev-upgrade-legacy.2
 - @slicemachine/adapter-sveltekit@0.3.17-dev-upgrade-legacy.2
 - @slicemachine/init@2.6.4-dev-upgrade-legacy.2
 - @slicemachine/manager@0.11.3-dev-upgrade-legacy.2
 - @slicemachine/plugin-kit@0.4.17-dev-upgrade-legacy.2
 - slice-machine-ui@1.14.1-dev-upgrade-legacy.2
 - start-slicemachine@0.11.7-dev-upgrade-legacy.2

lihbr committed:
Merge branch 'dev-next-release' of github.com:prismicio/slice-machine into lh/convert-legacy-slices

lihbr committed:
Publish

 - cimsirp@1.15.1-dev-upgrade-legacy.0
 - next-upgrade@1.14.1-dev-upgrade-legacy.3
 - sveltekit@1.15.1-dev-upgrade-legacy.0
 - @slicemachine/adapter-next@0.3.19-dev-upgrade-legacy.0
 - @slicemachine/adapter-nuxt@0.3.19-dev-upgrade-legacy.0
 - @slicemachine/adapter-nuxt2@0.3.19-dev-upgrade-legacy.0
 - @slicemachine/adapter-sveltekit@0.3.19-dev-upgrade-legacy.0
 - @slicemachine/init@2.7.2-dev-upgrade-legacy.0
 - @slicemachine/manager@0.12.1-dev-upgrade-legacy.0
 - @slicemachine/plugin-kit@0.4.19-dev-upgrade-legacy.0
 - slice-machine-ui@1.15.1-dev-upgrade-legacy.0
 - start-slicemachine@0.11.9-dev-upgrade-legacy.0

lihbr committed:
refactor: remove old store actions

lihbr committed:
Merge branch 'lh/convert-legacy-slices' of github.com:prismicio/slice-machine into lh/convert-legacy-slices

lihbr committed:
Publish

 - cimsirp@1.15.1-dev-upgrade-legacy.1
 - next-upgrade@1.14.1-dev-upgrade-legacy.4
 - sveltekit@1.15.1-dev-upgrade-legacy.1
 - @slicemachine/adapter-next@0.3.19-dev-upgrade-legacy.1
 - @slicemachine/adapter-nuxt@0.3.19-dev-upgrade-legacy.1
 - @slicemachine/adapter-nuxt2@0.3.19-dev-upgrade-legacy.1
 - @slicemachine/adapter-sveltekit@0.3.19-dev-upgrade-legacy.1
 - @slicemachine/init@2.7.2-dev-upgrade-legacy.1
 - @slicemachine/manager@0.12.1-dev-upgrade-legacy.1
 - @slicemachine/plugin-kit@0.4.19-dev-upgrade-legacy.1
 - slice-machine-ui@1.15.1-dev-upgrade-legacy.1
 - start-slicemachine@0.11.9-dev-upgrade-legacy.1

lihbr committed:
feat(manager,slice-machine): support legacy slices

lihbr committed:
Merge branch 'lh/convert-legacy-slices' of github.com:prismicio/slice-machine into lh/convert-legacy-slices

lihbr committed:
Publish

 - cimsirp@1.15.1-dev-upgrade-legacy.2
 - next-upgrade@1.14.1-dev-upgrade-legacy.5
 - sveltekit@1.15.1-dev-upgrade-legacy.2
 - @slicemachine/adapter-next@0.3.19-dev-upgrade-legacy.2
 - @slicemachine/adapter-nuxt@0.3.19-dev-upgrade-legacy.2
 - @slicemachine/adapter-nuxt2@0.3.19-dev-upgrade-legacy.2
 - @slicemachine/adapter-sveltekit@0.3.19-dev-upgrade-legacy.2
 - @slicemachine/init@2.7.2-dev-upgrade-legacy.2
 - @slicemachine/manager@0.12.1-dev-upgrade-legacy.2
 - @slicemachine/plugin-kit@0.4.19-dev-upgrade-legacy.2
 - slice-machine-ui@1.15.1-dev-upgrade-legacy.2
 - start-slicemachine@0.11.9-dev-upgrade-legacy.2

bapmrl commented on .github/workflows/ci.yml:
```diff
@@ -184,7 +184,7 @@ jobs:
   deployment:
     runs-on: ubuntu-latest
-    if: startsWith(github.ref, 'refs/heads/dev') # Job only runs on dev branches
+    if: startsWith(github.ref, 'refs/heads/lh') # Job only runs on dev branches
```
⚠️ I'm pretty sure you are aware of that, but the two changes to `.github/workflows/ci.yml` should be reverted to avoid breaking the `deployment` job.

lihbr committed:
feat: allow merging legacy slices

lihbr committed:
Publish

 - cimsirp@1.15.1-dev-upgrade-legacy.3
 - next-upgrade@1.14.1-dev-upgrade-legacy.6
 - sveltekit@1.15.1-dev-upgrade-legacy.3
 - @slicemachine/adapter-next@0.3.19-dev-upgrade-legacy.3
 - @slicemachine/adapter-nuxt@0.3.19-dev-upgrade-legacy.3
 - @slicemachine/adapter-nuxt2@0.3.19-dev-upgrade-legacy.3
 - @slicemachine/adapter-sveltekit@0.3.19-dev-upgrade-legacy.3
 - @slicemachine/init@2.7.2-dev-upgrade-legacy.3
 - @slicemachine/manager@0.12.1-dev-upgrade-legacy.3
 - @slicemachine/plugin-kit@0.4.19-dev-upgrade-legacy.3
 - slice-machine-ui@1.15.1-dev-upgrade-legacy.3
 - start-slicemachine@0.11.9-dev-upgrade-legacy.3

bapmrl commented on package.json:
```diff
@@ -12,7 +12,7 @@
     "dev:adapter-next": "yarn workspace @slicemachine/adapter-next dev",
     "dev:adapter-sveltekit": "yarn workspace @slicemachine/adapter-sveltekit dev",
-    "dev:e2e-next": "cd ./e2e-projects/next && yarn dev --port 8000",
+    "dev:e2e-next": "cd ./e2e-projects/next-upgrade && yarn dev --port 8000",
```
💡 Maybe this change should be reverted before merging as it won't be the most commonly used project during development?

bapmrl commented on e2e-projects/next-upgrade/package.json:
```diff
@@ -0,0 +1,28 @@
+    "next": "13.4.17",
+    "react": "18.2.0",
+    "react-dom": "18.2.0",
+    "typescript": "5.1.6"
```
❓ What do you think of ensuring each dependency has a single version within our monorepo? I think it would be easier to maintain (avoiding combinatorial issues) and easier to track what features we can use or not when we develop (for instance, why should we be able to use the new `satisfies` operator only in the new `e2e-projects/next-upgrade`?).

lihbr replied:
Aligned with what's on `slice-machine-ui`!

bapmrl commented on e2e-projects/next-upgrade/package.json:
```diff
@@ -0,0 +1,28 @@
+    "@types/node": "20.5.0",
+    "@types/react": "18.2.20",
+    "@types/react-dom": "18.2.7",
+    "next": "13.4.17",
```
❓ What do you think of using only `^` ranges for all `dependencies` and strict versions for all `devDependencies`? It's what's been done lately for consistency.

lihbr replied:
Updated

bapmrl commented on e2e-projects/next-upgrade/package.json:
```diff
@@ -0,0 +1,28 @@
+    "build": "next build",
+    "start": "next start",
+    "lint": "next lint",
+    "slicemachine": "start-slicemachine"
```
❓ Should we set the `NODE_ENV=development SM_ENV=staging` environment variables like in the other `e2e-projects/next` project?

lihbr replied:
Yes!

bapmrl commented on e2e-projects/next-upgrade/package.json:
```diff
@@ -0,0 +1,28 @@
+    "@prismicio/client": "^7.1.1",
+    "@prismicio/next": "^1.3.4",
+    "@prismicio/react": "^2.7.1",
+    "@types/node": "20.5.0",
```
💡 I would tend to move `@types/*` and other development specific dependencies to `devDependencies`, no?

lihbr replied:
It's what `npx create-next-app` created 🤷‍♀️

lihbr replied:
Reordered anyway

bapmrl commented on e2e-projects/next-upgrade/.gitignore:
```diff
@@ -0,0 +1,35 @@
+# See https://help.github.com/articles/ignoring-files/ for more about ignoring files.
```
💡 I'm okay to have specific configuration files in each workspace for now, but I think we should discuss what we would like to do in the future ^^ I think it's a lot of work to handle so many small variations in a monorepo.

lihbr replied:
I removed it, but I all fairness I was just rolling with the defaults of `npx create-next-app` to avoid overthinking it

bapmrl commented on e2e-projects/next-upgrade/tsconfig.json:
```diff
@@ -0,0 +1,27 @@
+{
```
❓ I thought Next.js was coming with its own TypeScript configuration and there was no need to define it in userland?

lihbr replied:
I guess it now wants one anyway, against, just plain default `npx create-next-app`

bapmrl commented on e2e-projects/next-upgrade/slicemachine.config.json:
```diff
@@ -0,0 +1,9 @@
+{
+  "___repositoryName": "inspiring-briouat-syt67g",
```
❓ Why is there multiple repository names? They should probably be removed before merging (or a comment should be added to explain the intent).

lihbr replied:
Cleaned up

bapmrl commented on e2e-projects/next-upgrade/app/slice-simulator/page.tsx:
```diff
@@ -0,0 +1,14 @@
+import { SliceZone } from "@prismicio/react";
+import { SliceSimulator } from "@slicemachine/adapter-next/simulator";
+
+import { components } from "../../slices";
```
⚠️ I couldn't find the `slices` folder. Is it intended?

lihbr replied:
Yes, there are no slices on this project by default, it's just a project legacy project after Slice Machine init :)

lihbr committed:
Merge branch 'master' of github.com:prismicio/slice-machine into lh/convert-legacy-slices

lihbr committed:
Merge branch 'lh/convert-legacy-slices' of github.com:prismicio/slice-machine into lh/convert-legacy-slices

lihbr committed:
Publish

 - cimsirp@1.16.1-dev-upgrade-legacy-2.0
 - next-upgrade@1.14.1-dev-upgrade-legacy-2.0
 - sveltekit@1.16.1-dev-upgrade-legacy-2.0
 - @slicemachine/adapter-next@0.3.20-dev-upgrade-legacy-2.0
 - @slicemachine/adapter-nuxt@0.3.20-dev-upgrade-legacy-2.0
 - @slicemachine/adapter-nuxt2@0.3.20-dev-upgrade-legacy-2.0
 - @slicemachine/adapter-sveltekit@0.3.20-dev-upgrade-legacy-2.0
 - @slicemachine/init@2.7.3-dev-upgrade-legacy-2.0
 - @slicemachine/manager@0.13.1-dev-upgrade-legacy-2.0
 - @slicemachine/plugin-kit@0.4.20-dev-upgrade-legacy-2.0
 - slice-machine-ui@1.16.1-dev-upgrade-legacy-2.0
 - start-slicemachine@0.11.10-dev-upgrade-legacy-2.0

bapmrl commented on e2e-projects/next-upgrade/customtypes/kitchen_sink/index.json:
```diff
@@ -0,0 +1,123 @@
+{
```
❓ What is the intent behind having these custom types? I didn't find any legacy slice inside them (maybe I should refresh my memory if I missed them :p). Are there some legacy slices in the remote Prismic repository?

lihbr replied:
All the slices in these custom types are legacy slices :)

lihbr committed:
style: fix linter

lihbr committed:
Merge branch 'lh/convert-legacy-slices' of github.com:prismicio/slice-machine into lh/convert-legacy-slices

lihbr committed:
Publish

 - cimsirp@1.16.1-dev-upgrade-legacy-2.1
 - next-upgrade@1.14.1-dev-upgrade-legacy-2.1
 - sveltekit@1.16.1-dev-upgrade-legacy-2.1
 - @slicemachine/adapter-next@0.3.20-dev-upgrade-legacy-2.1
 - @slicemachine/adapter-nuxt@0.3.20-dev-upgrade-legacy-2.1
 - @slicemachine/adapter-nuxt2@0.3.20-dev-upgrade-legacy-2.1
 - @slicemachine/adapter-sveltekit@0.3.20-dev-upgrade-legacy-2.1
 - @slicemachine/init@2.7.3-dev-upgrade-legacy-2.1
 - @slicemachine/manager@0.13.1-dev-upgrade-legacy-2.1
 - @slicemachine/plugin-kit@0.4.20-dev-upgrade-legacy-2.1
 - slice-machine-ui@1.16.1-dev-upgrade-legacy-2.1
 - start-slicemachine@0.11.10-dev-upgrade-legacy-2.1

bapmrl commented on packages/init/src/SliceMachineInitProcess.ts:
```diff
@@ -1285,7 +1289,41 @@ ${chalk.cyan("?")} Your Prismic repository name`.replace("\n", ""),
+								}
+
+								if (ids && ids.length > 0) {
+									task.skip("Types already exists");
```
⚠️ Typo: `Types already exist`.

lihbr committed:
fix(slice-machine): types

lihbr committed:
Publish

 - cimsirp@1.16.1-dev-upgrade-legacy-2.2
 - next-upgrade@1.14.1-dev-upgrade-legacy-2.2
 - sveltekit@1.16.1-dev-upgrade-legacy-2.2
 - @slicemachine/adapter-next@0.3.20-dev-upgrade-legacy-2.2
 - @slicemachine/adapter-nuxt@0.3.20-dev-upgrade-legacy-2.2
 - @slicemachine/adapter-nuxt2@0.3.20-dev-upgrade-legacy-2.2
 - @slicemachine/adapter-sveltekit@0.3.20-dev-upgrade-legacy-2.2
 - @slicemachine/init@2.7.3-dev-upgrade-legacy-2.2
 - @slicemachine/manager@0.13.1-dev-upgrade-legacy-2.2
 - @slicemachine/plugin-kit@0.4.20-dev-upgrade-legacy-2.2
 - slice-machine-ui@1.16.1-dev-upgrade-legacy-2.2
 - start-slicemachine@0.11.10-dev-upgrade-legacy-2.2

bapmrl commented on packages/init/test/SliceMachineInitProcess-syncDataWithPrismic.test.ts:
```diff
@@ -675,9 +701,74 @@ it("throws if context is missing repository", async (ctx) => {
+	expect(stdout).toMatch(/Pulled existing types/);
+});
+
+it("skips pulling types if types already exists locally", async (ctx) => {
```
⚠️ Typo: "types already exist".

bapmrl commented on packages/manager/src/managers/slices/SlicesManager.ts:
```diff
@@ -404,6 +429,107 @@ export class SlicesManager extends BaseManager {
+		const legacySliceAsVariation: Variation = {
+			id: args.dest.variationID,
+			name: args.dest.variationName,
+			description: args.dest.variationName,
```
❓ Does it make sense to use `variationName` as a `description`? Maybe it should be left empty?

lihbr replied:
That's to align with what has been done so far: https://github.com/prismicio/slice-machine/blob/8a68d99c7dd7a00042998198e73f29bf5aea2767/packages/slice-machine/src/features/slices/actions/createSlice.ts#L49

Let me know about it

bapmrl commented on packages/manager/src/managers/slices/SlicesManager.ts:
```diff
@@ -404,6 +429,107 @@ export class SlicesManager extends BaseManager {
+					id: args.dest.sliceID,
+					type: "SharedSlice",
+					name: args.dest.sliceID,
+					legacyPaths: {
```
❓ What is the purpose of `legacyPaths`?

lihbr replied:
It's a new property introduced in `@prismicio/types-internal` to allow the backend to resolve composite and legacy slices to the correct model: https://www.notion.so/prismic/RFC-Introducing-legacyPaths-to-Shared-Slices-in-Prismic-to-streamline-transition-to-Slice-Machine-8ccf067c394f430d867b861cb084ed32

bapmrl commented on packages/manager/src/managers/slices/SlicesManager.ts:
```diff
@@ -404,6 +429,107 @@ export class SlicesManager extends BaseManager {
+					...maybeExistingSlice.variations,
+					legacySliceAsVariation,
+				];
+			}
```
⚠️ I'm not sure I understand the `else` case. What happens when `maybeVariation` is truthy? Does it mean the slice as already been converted and we just change `legacyPaths`? Maybe a comment could help clarify the intent.

lihbr replied:
There are 3 cases:
1. Convert to new shared slice
2. Convert to a new variation
3. Merge with an existing and identical variation

This if covers cases n°2 and n°3, the if `if (!maybeVariation)` covers n°2 (because if the target variation does not exist, then it means we're in case n°2)

I updated comments, let me know

bapmrl commented on packages/slice-machine/src/features/slices/sliceCards/NonSharedSliceViewCard.tsx:
```diff
@@ -1,15 +1,23 @@
 type NonSharedSliceViewCardProps = {
-  slice: NonSharedSlice;
+  slice: NonSharedSliceInSliceZone;
+  path: {
```
💡 This type is also defined inline in `packages/slice-machine/components/Forms/ConvertLegacySliceModal/types.ts`. Maybe it could be extracted to its own named typed and imported from there?

lihbr replied:
Refactored!

bapmrl commented on packages/slice-machine/components/Forms/ConvertLegacySliceModal/index.tsx:
```diff
@@ -0,0 +1 @@
+export { ConvertLegacySliceModal } from "./ConvertLegacySliceModal";
```
⚠️ We would like to put new code in `src/features/*` (and in this case `src/features/slices/???`).

lihbr replied:
Moved it to `src/features/slices/convertLegacySlice`, let me know if you prefer another location!

bapmrl commented on packages/slice-machine/src/features/slices/sliceCards/NonSharedSliceViewCard.tsx:
```diff
@@ -21,12 +29,16 @@ export const NonSharedSliceViewCard: FC<NonSharedSliceViewCardProps> = ({
       >
         <Badge color="purple" title="Legacy Slice" />
       </Tooltip>
+      <ConvertLegacySliceModal slice={slice} path={path} />
```
⚠️ I think the name of that component is a bit misleading. The first thing it renders is a `Button` than itself triggers a `DropdownMenu`. I think `ConvertLegacySliceButton` would better reflect what it is.

lihbr replied:
Refactored!

bapmrl commented on packages/slice-machine/lib/builders/CustomTypeBuilder/SliceZone/index.tsx:
```diff
@@ -76,7 +76,7 @@ const mapAvailableAndSharedSlices = (
       }
       // Legacy Slice
-      else if (value.type === "Slice") {
+      else if (typeof value.type === "string") {
```
💡 Why not just `else` instead of `else if`? Also, the comments seem to be outdated.

lihbr replied:
Updated the comment and refactored to remove the unnecessary `else if`

bapmrl commented on packages/slice-machine/lib/builders/CustomTypeBuilder/SliceZone/index.tsx:
```diff
@@ -154,6 +154,14 @@ const SliceZone: React.FC<SliceZoneProps> = ({
     .map((e) => e.payload) as ReadonlyArray<ComponentUI>;
 
+  /* Preserve these keys in SliceZone */
+  const path = useMemo(() => {
```
❓ Is this `useMemo` important to make the code work or is it an optimization?

lihbr replied:
I guess it's just an optimization, I'm not familiar with how reactivity works in React, would the following also be reactive?

```tsx
const path = {
	customTypeID: customType.id,
	tabID: tabId,
	sliceZoneID: sliceZone?.key as string,
};
```

lihbr replied:
Refactored!

bapmrl commented on packages/slice-machine/lib/builders/CustomTypeBuilder/SliceZone/List.tsx:
```diff
@@ -16,12 +16,18 @@ import { SharedSliceViewCard } from "@src/features/slices/sliceCards/SharedSlice
 interface SlicesListProps {
   format: CustomTypeFormat;
   slices: ReadonlyArray<SliceZoneSlice>;
+  path: {
```
💡 I think this could be a shared type.

lihbr replied:
Where could it be extracted in your opinion?

bapmrl commented on packages/slice-machine/components/Forms/ConvertLegacySliceModal/ConvertLegacySliceModal.tsx:
```diff
@@ -0,0 +1,295 @@
+  Type,
+} from "./types";
+
+const getFieldMappingFingerprint = (
```
⚠️ I think this file has too many responsibilities and is a bit too long. For instance, this function could probably be extracted somewhere else, no? I've created a `src/domain/slice.ts` file to contain shared domain logic. Do you think it would be a good fit?

lihbr replied:
Definitely!

bapmrl commented on packages/slice-machine/components/Forms/ConvertLegacySliceModal/ConvertLegacySliceModal.tsx:
```diff
@@ -0,0 +1,295 @@
+  primary: string;
+  items: string;
+} => {
+  const primary: Record<string, string> = {};
```
💡 I know `@prismicio/types-internal` uses a lot of `Record`, but this type is actually false because there exists some `string` keys that don't have a corresponding `string` value, no? `Partial<Record<string, string>>` or `Map<string, string>` would better convey this.

lihbr replied:
Type is always defined here, that's why we're using `?? ""` later on. Everything gets stringified before the function returns anyway to get a fingerprint, so it looks like it's pretty scoped here?

`Map` would be too overkill to deal with I think~

bapmrl commented on packages/slice-machine/components/Forms/ConvertLegacySliceModal/ConvertLegacySliceModal.tsx:
```diff
@@ -0,0 +1,295 @@
+  const localSharedSlices = useMemo(() => {
+    return libraries.map((library) => library.components).flat();
+  }, [libraries]);
+  const identicalSlices = useMemo<IdenticalSlice[]>(() => {
```
⚠️ Moving this code block into a `useIdenticalSlices` hook (in the same file or in another file) would make `ConvertLegacySliceModal` easier to read.

lihbr replied:
I tried refactoring it, let me know if that reads better :)

bapmrl commented on packages/slice-machine/components/Forms/ConvertLegacySliceModal/ConvertLegacySliceModal.tsx:
```diff
@@ -0,0 +1,295 @@
+  } = useSliceMachineActions();
+
+  const [isLoading, setIsLoading] = useState(false);
+  const [isModalOpen, setIsModalOpen] = useState<false | Type>(false);
```
💡 "Modal" isn't a good name for me, because it only describes a UX pattern where users can't click/interact outside of the element. Also, "isOpen" makes me think it's a boolean, but in fact you want to know which modal is open. Something like `const [dialog, setDialog] = useState<LegacySliceConversionType | undefined>()` would be clearer in my opinion (or maybe `ConvertLegacySliceDialogType` instead of `LegacySliceConversionType`).

lihbr replied:
Refactored!

bapmrl commented on packages/slice-machine/components/Forms/ConvertLegacySliceModal/ConvertLegacySliceModal.tsx:
```diff
@@ -0,0 +1,295 @@
+    }
+
+    setIsLoading(true);
+    void (async () => {
```
💡 Is this IIFE necessary? It's not super easy to read in my opinion.

lihbr replied:
Refactored, that was because the submit function couldn't be async

bapmrl commented on packages/slice-machine/components/Forms/ConvertLegacySliceModal/ConvertLegacySliceModal.tsx:
```diff
@@ -0,0 +1,295 @@
+    })
+  );
+
+  const {} = useSliceMachineActions();
```
⚠️ I don't understand why we call this hook as the return type isn't used.

lihbr replied:
Cleaned up 🙈

bapmrl commented on packages/slice-machine/components/Forms/ConvertLegacySliceModal/ConvertLegacySliceModal.tsx:
```diff
@@ -0,0 +1,295 @@
+        });
+
+      if (errors.length) {
+        console.error(
```
❓ Do we want to show a toast in case the operation doesn't work?

lihbr replied:
Added a toast upon conversion error and updating the UI after 👌

bapmrl commented on packages/slice-machine/components/Forms/ConvertLegacySliceModal/ConvertLegacySliceModal.tsx:
```diff
@@ -0,0 +1,295 @@
+          break;
+      }
+
+      const customTypeSM = CustomTypes.toSM(customType);
```
❓ Does it work if called immediately after `refreshState(serverState)`?

lihbr replied:
Yes, it's synchronous so I think it's all good?

lihbr replied:
Good as is

bapmrl commented on packages/slice-machine/components/Forms/ConvertLegacySliceModal/ConvertLegacySliceModal.css.ts:
```diff
@@ -0,0 +1,35 @@
+import { sprinkles } from "@prismicio/editor-ui";
+import { style } from "@vanilla-extract/css";
+
+export const form = style([
```
💡 You don't have to change your styles, but next time I would advice you to use `Box`. Inline with the Page Builder's team philosophy, we now want to:
1. Write new components with Storybook in `src/components` when they are specific to Slice Machine. Write them in the Editor's repository when they are sharable. 
2. Avoid writing styles with `@vanilla-extract/*` in domain code. Fallback to `Box` for contextual styles/layouts.

lihbr replied:
Ohhh, ok, now I get `Box`

Refactored what I could with it :)

bapmrl commented on packages/slice-machine/components/Forms/ConvertLegacySliceModal/ConvertLegacySliceModal.css.ts:
```diff
@@ -0,0 +1,35 @@
+import { style } from "@vanilla-extract/css";
+
+export const form = style([
+  sprinkles({
```
⚠️ No need to wrap `sprinkles` in `style`. You can directly call `sprinkles` (e.g., `export const form = sprinkles({ /* ... */ })`).

lihbr replied:
Refactored!

bapmrl commented on packages/slice-machine/components/Forms/ConvertLegacySliceModal/types.ts:
```diff
@@ -0,0 +1,48 @@
+import { LibraryUI } from "@models/common/LibraryUI";
+import { ComponentUI } from "@models/common/ComponentUI";
+
+export type ConvertLegacySliceModalProps = {
```
⚠️ For the sake of consistency, I think it's better to have props defined in the same file as the component using them. However, when props are shared, like your `FormProps`, I think it's fine to have them in a `types` file.

lihbr replied:
Refactored :)

bapmrl commented on packages/slice-machine/components/Forms/ConvertLegacySliceModal/types.ts:
```diff
@@ -0,0 +1,48 @@
+  "as_new_variation",
+  "merge_with_identical",
+] as const;
+export type Type = (typeof types)[number];
```
⚠️ This is a too generic name 😊 What do you think of `LegacySliceConversionType` (or maybe `ConvertLegacySliceDialogType`)?

lihbr replied:
Yes! Refactored

bapmrl commented on packages/slice-machine/components/Forms/ConvertLegacySliceModal/types.ts:
```diff
@@ -0,0 +1,48 @@
+  };
+};
+
+export type ConvertLegacySliceAndTrackArgs = {
```
⚠️ For the sake of consistency, in the front end we tend to define args and props close to the function or component using them.

lihbr replied:
I agree, however, since it needs to be shared, I'm a bit clueless as of what's better to not have a circular dependency 🤔

bapmrl commented on packages/slice-machine/components/Forms/ConvertLegacySliceModal/FormAsNewSlice.tsx:
```diff
@@ -0,0 +1,153 @@
+import * as styles from "./ConvertLegacySliceModal.css";
+import { FormProps } from "./types";
+
+export const FormAsNewSlice: React.FC<FormProps> = ({
```
💡 The uppermost component is a `Dialog` so I think `AsNewSliceDialog` would make more sense (or maybe `ConvertLegacySliceAsNewSliceDialog`). What do you think?

lihbr replied:
Refactored them all :)

bapmrl commented on packages/slice-machine/components/Forms/ConvertLegacySliceModal/FormAsNewSlice.tsx:
```diff
@@ -0,0 +1,153 @@
+          }}
+        >
+          {(formik) => {
+            // Just a hack to trigger validation on init for better UX
```
❓ https://formik.org/docs/api/formik#validateonmount-boolean doesn't work?

lihbr replied:
Ah, it does! Tried looking for it but didn't get to this page!

bapmrl commented on packages/slice-machine/components/Forms/ConvertLegacySliceModal/FormAsNewSlice.tsx:
```diff
@@ -0,0 +1,153 @@
+
+            return (
+              <form
+                id="convert-legacy-slice-modal-as-new-slice"
```
⚠️ You can remove them if you haven't written Cypress/Vitest UI tests ^^

lihbr replied:
Removed!

bapmrl commented on packages/slice-machine/components/Forms/ConvertLegacySliceModal/FormAsNewVariation.tsx:
```diff
@@ -0,0 +1,234 @@
+  );
+  // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
+  if (!errors.libraryID && !library) {
+    errors.libraryID = "Does not exists.";
```
⚠️ Typo: "Does not exist." (without an "s").

lihbr replied:
Fixed! 🙏

bapmrl commented on packages/slice-machine/components/Forms/ConvertLegacySliceModal/FormAsNewVariation.tsx:
```diff
@@ -0,0 +1,234 @@
+  return errors;
+};
+
+export const FormAsNewVariation: React.FC<FormProps> = ({
```
💡 Personally I think it's easier to read files when the most important export appears first ^^ I would move `FormAsNewVariation ` to the top and just follow "natural reading order" from top to bottom (I don't know if I'm clear).

lihbr replied:
I actually expect the main export to be last haha (since back in the days, we were doing `module.exports = { ... }` at the end of our JS files.

Let me know what would work best for you :)

lihbr replied:
Refactored!

bapmrl reviewed (changes requested):
Excellent work 😍

Don't worry, I've added 40 comments and I know it's a lot... I don't want you to make a heart attack and a lot of them are just questions. I've tried to put priorities using the following icons: ⚠️, 💡 and ❓.

Let's have a call to quickly go through the comments if you want 😊

bapmrl reviewed (commented):
Excellent work 😍

Don't worry, I've added 40 comments and I know it's a lot... I don't want you to make a heart attack and a lot of them are just questions. I've tried to put priorities using the following icons: ⚠️, 💡 and ❓.

Let's have a call to quickly go through the comments if you want 😊

lihbr committed:
refactor(e2e-projects): review simplification

lihbr committed:
refactor(manager): review

lihbr committed:
refactor(slice-machine): review

lihbr committed:
Publish

 - cimsirp@1.16.1-dev-upgrade-legacy-2.3
 - next-upgrade@1.14.1-dev-upgrade-legacy-2.3
 - sveltekit@1.16.1-dev-upgrade-legacy-2.3
 - @slicemachine/adapter-next@0.3.20-dev-upgrade-legacy-2.3
 - @slicemachine/adapter-nuxt@0.3.20-dev-upgrade-legacy-2.3
 - @slicemachine/adapter-nuxt2@0.3.20-dev-upgrade-legacy-2.3
 - @slicemachine/adapter-sveltekit@0.3.20-dev-upgrade-legacy-2.3
 - @slicemachine/init@2.7.3-dev-upgrade-legacy-2.3
 - @slicemachine/manager@0.13.1-dev-upgrade-legacy-2.3
 - @slicemachine/plugin-kit@0.4.20-dev-upgrade-legacy-2.3
 - slice-machine-ui@1.16.1-dev-upgrade-legacy-2.3
 - start-slicemachine@0.11.10-dev-upgrade-legacy-2.3

lihbr committed:
refactor(slice-machine): review

lihbr committed:
Merge branch 'lh/convert-legacy-slices' of github.com:prismicio/slice-machine into lh/convert-legacy-slices

lihbr committed:
Publish

 - cimsirp@1.16.1-dev-upgrade-legacy-2.4
 - next-upgrade@1.14.1-dev-upgrade-legacy-2.4
 - sveltekit@1.16.1-dev-upgrade-legacy-2.4
 - @slicemachine/adapter-next@0.3.20-dev-upgrade-legacy-2.4
 - @slicemachine/adapter-nuxt@0.3.20-dev-upgrade-legacy-2.4
 - @slicemachine/adapter-nuxt2@0.3.20-dev-upgrade-legacy-2.4
 - @slicemachine/adapter-sveltekit@0.3.20-dev-upgrade-legacy-2.4
 - @slicemachine/init@2.7.3-dev-upgrade-legacy-2.4
 - @slicemachine/manager@0.13.1-dev-upgrade-legacy-2.4
 - @slicemachine/plugin-kit@0.4.20-dev-upgrade-legacy-2.4
 - slice-machine-ui@1.16.1-dev-upgrade-legacy-2.4
 - start-slicemachine@0.11.10-dev-upgrade-legacy-2.4

lihbr committed:
ci: update workflow

lihbr committed:
Publish

 - cimsirp@1.16.1-dev-upgrade-legacy-3.0
 - next-upgrade@1.14.1-dev-upgrade-legacy-3.0
 - sveltekit@1.16.1-dev-upgrade-legacy-3.0
 - @slicemachine/adapter-next@0.3.20-dev-upgrade-legacy-3.0
 - @slicemachine/adapter-nuxt@0.3.20-dev-upgrade-legacy-3.0
 - @slicemachine/adapter-nuxt2@0.3.20-dev-upgrade-legacy-3.0
 - @slicemachine/adapter-sveltekit@0.3.20-dev-upgrade-legacy-3.0
 - @slicemachine/init@2.7.3-dev-upgrade-legacy-3.0
 - @slicemachine/manager@0.13.1-dev-upgrade-legacy-3.0
 - @slicemachine/plugin-kit@0.4.20-dev-upgrade-legacy-3.0
 - slice-machine-ui@1.16.1-dev-upgrade-legacy-3.0
 - start-slicemachine@0.11.10-dev-upgrade-legacy-3.0

bapmrl commented on packages/slice-machine/src/domain/slice.ts:
```diff
@@ -30,3 +30,51 @@ export function getScreenshotUrl(
 }
+
+export const getFieldMappingFingerprint = (
+  slice: LegacySlice | CompositeSlice | VariationSM,
```
💡 You could remove the `NonSharedSlice` type which is no longer used outside this module (or use it in your arg type) and for consistency, write a `function` instead of a `const`.

bapmrl commented on packages/slice-machine/src/features/slices/convertLegacySlice/ConvertLegacySliceButton.tsx:
```diff
@@ -0,0 +1,267 @@
+
+type ConvertLegacySliceButtonProps = NonSharedSliceViewCardProps;
+
+export const ConvertLegacySliceButton: React.FC<
```
💡 For consistency, we don't use the `React` namespace and import each type/value manually (e.g., `import type { FC } from "react`).

bapmrl commented on packages/slice-machine/src/features/slices/convertLegacySlice/ConvertLegacySliceButton.tsx:
```diff
@@ -0,0 +1,267 @@
+  );
+
+  const sliceName =
+    slice.value.type === "Slice"
```
💡 I don't know if it's worth it and how we could name it, but that's the kind of domain function I would move to `packages/slice-machine/src/domain/slice.ts`.

lihbr replied:
I'm not really sure about one for now so I'll leave it as is 🤔

bapmrl replied:
Okay!

bapmrl reviewed (approved):
Excellent work Lucie!

I've just added 3 minor comments, that don't necessarily need to be addressed.

lihbr committed:
refactor(slice-machine): type consistency

lihbr committed:
Publish

 - cimsirp@1.16.1-dev-upgrade-legacy-3.1
 - next-upgrade@1.14.1-dev-upgrade-legacy-3.1
 - sveltekit@1.16.1-dev-upgrade-legacy-3.1
 - @slicemachine/adapter-next@0.3.20-dev-upgrade-legacy-3.1
 - @slicemachine/adapter-nuxt@0.3.20-dev-upgrade-legacy-3.1
 - @slicemachine/adapter-nuxt2@0.3.20-dev-upgrade-legacy-3.1
 - @slicemachine/adapter-sveltekit@0.3.20-dev-upgrade-legacy-3.1
 - @slicemachine/init@2.7.3-dev-upgrade-legacy-3.1
 - @slicemachine/manager@0.13.1-dev-upgrade-legacy-3.1
 - @slicemachine/plugin-kit@0.4.20-dev-upgrade-legacy-3.1
 - slice-machine-ui@1.16.1-dev-upgrade-legacy-3.1
 - start-slicemachine@0.11.10-dev-upgrade-legacy-3.1

bapmrl committed:
Merge branch 'dev-next-release' into lh/convert-legacy-slices

bapmrl committed:
revert ci.yml change
